File: blue/src/colors_compute/compute_options.py
"""Optional reviewed VM fields; omitted requirements preserve provider defaults."""
from importlib.resources import files
import json
from ._copy import deepcopy

def apply_options(provider, stage, request, documents, opts=None):
    descriptor = json.loads(files('colors_compute').joinpath('compute-options.json').read_text()).get(provider)
    request = deepcopy(request)
    option = (descriptor or {}).get('backups_option')
    if 'backups' not in request and option in (opts or {}):
        request['backups'] = {'enabled': opts[option]}
    selected = {key: request[key] for key in ('backups', 'ipv6') if key in request}
    if not selected:
        return documents
    if descriptor is None:
        raise ValueError('unsupported compute options capability')
    if 'ipv6' in selected and 'ipv6_field' not in descriptor:
        raise ValueError('unsupported compute options capability')
    if 'ipv6' in selected and type(selected['ipv6']) is not bool:
        raise ValueError('invalid compute IPv6 policy')
    backup = selected.get('backups')
    if 'backups' in selected:
        if not isinstance(backup, dict) or type(backup.get('enabled')) is not bool or set(backup) != ({'enabled', 'schedule'} if backup['enabled'] and 'schedule_field' in descriptor else {'enabled'}):
            raise ValueError('invalid compute backup policy')
        if backup['enabled'] and 'schedule_field' in descriptor:
            schedule = backup['schedule']
            if not isinstance(schedule, dict) or set(schedule) != {'type', 'hour'} or schedule['type'] != 'daily' or type(schedule['hour']) not in (int,float) or not 0 <= schedule['hour'] <= 23 or schedule['hour'] != int(schedule['hour']):
                raise ValueError('invalid compute backup schedule')
    if stage != 'node':
        return documents
    result = deepcopy(documents)
    resources = [doc.get('resource', {}).get(descriptor['resource_type'], {}).get(descriptor['resource_name']) for doc in result.values()]
    resources = [resource for resource in resources if resource is not None]
    if len(resources) != 1:
        raise ValueError('compute options resource unavailable')
    resource = resources[0]
    if 'ipv6' in selected:
        resource[descriptor['ipv6_field']] = selected['ipv6']
    if backup is not None:
        resource[descriptor['backups_field']] = descriptor['backups_values'][str(backup['enabled']).lower()]
        if backup['enabled'] and 'schedule_field' in descriptor:
            resource[descriptor['schedule_field']] = [{'type':'daily','hour':int(backup['schedule']['hour'])}]
    return result

File: red/src/compute-options.ts
import descriptors from '../resources/compute-options.json';
type Map=Record<string,any>;
const object=(v:any):v is Map=>v!==null&&typeof v==='object'&&!Array.isArray(v);
export function applyOptions(provider:string,stage:string,request:Map,documents:Map,opts:Map={}):Map {
 const descriptor=(descriptors as Map)[provider];request=structuredClone(request);if(!Object.hasOwn(request,'backups')&&descriptor?.backups_option&&Object.hasOwn(opts,descriptor.backups_option))request.backups={enabled:opts[descriptor.backups_option]};
 const keys=['backups','ipv6'].filter(k=>Object.hasOwn(request,k));if(!keys.length)return documents;
 if(!Object.hasOwn(descriptors,provider))throw Error('unsupported compute options capability');
 if(keys.includes('ipv6')&&!descriptor.ipv6_field)throw Error('unsupported compute options capability');
 if(keys.includes('ipv6')&&typeof request.ipv6!=='boolean')throw Error('invalid compute IPv6 policy');
 const backup=request.backups;
 if(keys.includes('backups')){if(!object(backup)||typeof backup.enabled!=='boolean'||Object.keys(backup).sort().join(',')!==(backup.enabled&&descriptor.schedule_field?'enabled,schedule':'enabled'))throw Error('invalid compute backup policy');if(backup.enabled&&descriptor.schedule_field){const schedule=backup.schedule;if(!object(schedule)||Object.keys(schedule).sort().join(',')!=='hour,type'||schedule.type!=='daily'||!Number.isSafeInteger(schedule.hour)||schedule.hour<0||schedule.hour>23)throw Error('invalid compute backup schedule');}}
 if(stage!=='node')return documents;const result=structuredClone(documents),resources=Object.values(result).map(doc=>doc.resource?.[descriptor.resource_type]?.[descriptor.resource_name]).filter(v=>v!==undefined);if(resources.length!==1)throw Error('compute options resource unavailable');const resource=resources[0];
 if(keys.includes('ipv6'))resource[descriptor.ipv6_field]=request.ipv6;if(backup!==undefined){resource[descriptor.backups_field]=descriptor.backups_values[String(backup.enabled)];if(backup.enabled&&descriptor.schedule_field)resource[descriptor.schedule_field]=[{type:'daily',hour:backup.schedule.hour}];}return result;
}
